merkle/compact/tree.go

```diff
@@ -129,33 +133,31 @@ func (t *Tree) CurrentRoot() []byte {
 
 type setNodeFunc func(depth int, index int64, hash []byte) error
 
+// recalculateRoot updates the current root hash, and calls setNodeFunc for
+// imperfect subtrees along the right border of the tree.
+//
+// TODO(pavelkalinnikov): Run this only once, after multiple AddLeaf calls.
 func (t *Tree) recalculateRoot(f setNodeFunc) error {
-	if t.size == 0 {
+	hashes := t.rng.Hashes()
+	ln, size := len(hashes), t.rng.End()
+	if size == 0 {
 		return nil
 	}
+	if got, want := bits.OnesCount64(size), ln; got != want {
+		return fmt.Errorf("hashes size mismatch: got %d, want %d", got, want)
+	}
 
-	index := t.size
-
-	var newRoot []byte
-	first := true
-	mask := int64(1)
-	numBits := bits.Len64(uint64(t.size))
-	for bit := 0; bit < numBits; bit++ {
-		index >>= 1
-		if t.size&mask != 0 {
-			if first {
-				newRoot = t.nodes[bit]
-				first = false
-			} else {
-				newRoot = t.hasher.HashChildren(t.nodes[bit], newRoot)
-				if err := f(bit+1, index, newRoot); err != nil {
-					return err
-				}
-			}
+	hash := hashes[ln-1]
+	for i := ln - 2; i >= 0; i-- {
+		size &= size - 1
+		hash = t.hasher.HashChildren(hashes[i], hash)
+		level := bits.TrailingZeros64(size) + 1
+		index := size >> uint(level)
```

Comment:
Can you document this section of code a bit more. On first reading it looks like there should be a `-1' here.